[git-webkit revert] Commit message should present more useful information

briannafan opened the pull request:
#### cab7e2d834d95d64c75955f7d66e154f262b52ad
<pre>
[git-webkit revert] Commit message should present more useful information
<a href="https://bugs.webkit.org/show_bug.cgi?id=263456">https://bugs.webkit.org/show_bug.cgi?id=263456</a>
<a href="https://rdar.apple.com/117270674">rdar://117270674</a>

Reviewed by Jonathan Bedard.

Commit message is formatted after webkit patch workflow.
Changed git-webkit revert logic to support these changes and future work.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.parser):
(Revert.get_issue_info):
(Revert):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:

Canonical link: <a href="https://commits.webkit.org/270930@main">https://commits.webkit.org/270930@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b2b4bd79713089ef2434ea8230a9ac2f554105d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5425 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28045 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29020 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24503 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27255 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2816 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24396 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27071 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23008 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3719 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26969 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3789 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29500 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24473 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24406 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30010 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3798 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1991 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27918 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/26581 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5250 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6451 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4265 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -316,11 +342,14 @@ def main(file_name=None, source=None, sha=None):
                 ['git', 'log', sha, '-1', '--pretty=format:%B'],
                 **ENCODING_KWARGS
             ))
+        elif commit_title.startswith('Unreviewed, reverting'):
```
This means that you won't trigger your revert function when someone runs a bare `git revert`. I think that's where we should start. If you rip out your `.git/hooks/prepare-commit-msg` hook and then run a `git revert`, you can see what a "raw" revert message is. I think we want to do something similar to what we do with `CHERRY_PICKING_RE` and `GIT_REFLOG_ACTION` action a few lines above this. I think I _tried_ to do something like that with `if source == 'merge' and not content.startswith('Revert'):`, but it doesn't look right, so you probably need to remove that line.

briannafan replied:
Landed a fix at #20127 to set `git revert` message to its default. Future work will make that message closer in format to `git webkit revert`, but that will come after fixing `git webkit revert` first.

JonWBedard commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -290,6 +288,34 @@ def cherry_pick(content):
+
+{content}
+'''.format(
+        title=os.environ.get('COMMIT_MESSAGE_TITLE', '') or 'Need a short description (OOPS!).',
```
You should probably have a default here that basically amounts to "Revert <commit ref>"

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -1222,7 +1222,7 @@ def clean(self):
         if staged in [True, False]:
             command = run(
-                [self.executable(), 'diff', '--name-only'] + (['--staged'] if staged else []),
+                [self.executable(), 'diff', '--name-only', '-v'] + (['--staged'] if staged else []),
```
Why are you adding a `-v` here?

briannafan replied:
Accidental artifact from debugging oops

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -30,17 +30,17 @@
 from ..commit import Commit
 
 
 class Revert(Command):
```
I think the changes to this class should be separated from the `prepare-commit-msg` changes. It may make sense to develop them as a stacked PR for now, but I think they should ultimately land separately. My reasoning for this is that I think that we should make an effort to make `git revert` do something mostly correct. While we can certainly encourage usage of `git-webkit revert`, I know there are some workflows where contributors will use `git revert`, and I'd like to be able to standardize those commits as well.

JonWBedard commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -121,7 +120,16 @@ def message(source=None, sha=None):
+    revert_msg = os.environ.get('COMMIT_MESSAGE_REVERT', '')
+    if revert_msg:
+        return '''{title}\n{revert}\n{bugs}
+    '''.format(
```
I think order is wrong here. Why are `bugs` coming after the revert message? Also, you're already using `\n` in this string, I think this string should be a single line.

briannafan replied:
It's because the `get_bugs_string()` is going to get the bugs for the commit that is being reverted, not the new revert bug. So that one needs to go under the "Reverted changeset:" line which is set in `revert`.

JonWBedard replied:
Got it!

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -59,11 +66,34 @@ def parser(cls, parser, loggers=None):
-        if repository.modified():
-            sys.stderr.write('Please commit your changes or stash them before you revert commit: {}'.format(args.commit))
-            return 1
+    def get_issue_info(cls, args, repository, **kwargs):
```
This is directly lifted from `program/branch.py`, yes? Have we looked at combining the logic?

briannafan replied:
In regard to consolidating with `branch.py`, there's a problem with using the existing infra with branchpoint/branch/PR creation:

We need to set an issue before calling PullRequest.pull_request_branch_point or else that code will set args.issue from the commit we want to revert and create a PR using that.

Right now the issue logic in `branch.py` is spread over its `main()`. So maybe the best solution is to throw all that in a function and then call it in `revert.py`. (As discussed, will be doing that in a separate change.)

briannafan committed:
[git-webkit revert] Commit message should present more useful information
https://bugs.webkit.org/show_bug.cgi?id=263456
rdar://117270674

Reviewed by Jonathan Bedard.

Commit message is formatted after webkit patch workflow.
Changed git-webkit revert logic to support these changes and future work.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.parser):
(Revert.get_issue_info):
(Revert):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:

Canonical link: https://commits.webkit.org/270930@main
